extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java

```diff
@@ -212,41 +213,191 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
-                service.classAnnotation(NON_BLOCKING) == null;
+        return collected;
+    }
 
-        for (MethodInfo method : service.methods()) {
-            if (BLOCKING_SKIPPED_METHODS.contains(method.name())) {
-                continue;
+    private enum BlockingMode {
+        UNDEFINED(false, false),
+        BLOCKING(true, true),
+        NON_BLOCKING(false, true),
+        TRANSACTIONAL(true, false);
+
+        private final boolean blocking;
+        private final boolean explicit;
+
+        BlockingMode(boolean blocking, boolean explicit) {
+            this.blocking = blocking;
+            this.explicit = explicit;
+        }
+    }
+
+    private static BlockingMode blockingMode(Predicate<DotName> checker, Supplier<String> exceptionMsgSupplier) {
+        boolean blocking = checker.test(BLOCKING);
+        boolean nonBlocking = checker.test(NON_BLOCKING);
+        if (blocking && nonBlocking) {
+            throw new DeploymentException(exceptionMsgSupplier.get());
+        }
+        if (blocking) {
+            return BlockingMode.BLOCKING;
+        }
+        if (nonBlocking) {
+            return BlockingMode.NON_BLOCKING;
+        }
+        boolean transactional = checker.test(TRANSACTIONAL);
+        if (transactional) {
+            return BlockingMode.TRANSACTIONAL;
+        }
+        return BlockingMode.UNDEFINED;
+    }
+
+    /**
+     * Retrieve the blocking-mode for the given method (by name + parameter types).
+     *
+     * <p>
+     * Traverses the service impl class hierarchy, stops at the first "explict" annotation
+     * ({@link io.smallrye.common.annotation.Blocking} or {@link io.smallrye.common.annotation.NonBlocking}).
+     *
+     * <p>
+     * Otherwise returns the "topmost" "non-explicit" annotation (aka {@link javax.transaction.Transactional}).
+     */
+    private static BlockingMode serviceBlockingMode(List<ClassInfo> classes) {
+        BlockingMode mode = BlockingMode.UNDEFINED;
+        for (ClassInfo ci : classes) {
+            BlockingMode classMode = blockingMode(annotationName -> ci.classAnnotation(annotationName) != null,
+                    () -> "Class '" + ci.name() + "' contains both @Blocking and @NonBlocking annotations.");
+
+            if (classMode.explicit) {
+                return classMode;
+            }
+            if (classMode != BlockingMode.UNDEFINED) {
+                mode = classMode;
             }
+        }
+        return mode;
+    }
 
-            if (method.hasAnnotation(BLOCKING)) {
-                if (method.hasAnnotation(NON_BLOCKING)) {
-                    throw new DeploymentException("Method '" + method.declaringClass().name() + "#" + method.name() +
-                            "' contains both @Blocking and @NonBlocking annotations.");
-                } else {
-                    result.add(method);
+    /**
+     * Retrieve the blocking-mode for the given method (by name + parameter types).
```

Comment:
@michalszynkiewicz I thought that gRPC method names must be unique?

So why the usage of param types here?